[Docs]: Hello world tutorial revamp (#1077)
risc0/risc0: website/docs/tutorials/hello-world.md

# Building zkVM Hello World

This tutorial will walk you through building your first zkVM application.
By following the steps in this guide, you will learn how:

- to use the [cargo risczero] tool to create a new project
- to send private data to the guest program

> - zkVM executes and generates a proof of a guest program
> - guest writes public output

Reviewer: Copyediting

```suggestion
- the zkVM executes and generates a proof of a guest program
- the guest writes public output
```

## Step 1: Create a new project

> Firstly, visit [Installation] page for installing necessary software.

Reviewer 2: ```suggestion
Firstly, visit the [installation] page for how to install all the necessary software.
```

> Secondly, using `risczero` toolchain create a `hello-world` project from the command line. The `risczero` toolchain takes `--guest-name` parameter, a guest program you want to run on zkVM to generate a proof of its execution:

Reviewer 2: ```suggestion
Secondly, using the `cargo-risczero` tool create a `hello-world` project from the command line. The `cargo-risczero` tool allows for a `--guest-name` parameter, a guest program you want to run on zkVM to generate a proof of its execution:
```

```bash
## Create a project from our starter template
cargo risczero new hello-world --guest-name hello_guest
cd hello-world
```

In the project folder, `hello-world`, build and run the project using `cargo run --release`.
Use this command any time you'd like to check your progress.

## Step 2 (Host): Share private data as input with the guest

zkVM or a [prover] runs on the **[host]**. The host code is in `hello-world/host/src/main.rs`.
The host creates an executor environment `ExecutorEnv` before constructing a prover.
The host makes `input` value available to the guest before execution. It does it by adding `input` to the executor environment, which is responsible for managing guest-readable memory. When the prover executes the program, it can access input:

> ```no_compile

Reviewer 2: It'd be great if we could get test coverage for this.

Author: @flaub I had to put `ignore` back because `examples/Hello World` is renamed `factorials` example. I'll create a ticket to update the `hello-world` example and fix this tutorial after Istanbul

> The host makes `input` value available to the guest before execution. It does it by adding `input` to the executor environment, which is responsible for managing guest-readable memory. When the prover executes the program, it can access input:

Reviewer: Copyediting
```suggestion
The host makes the value `input` available to the guest before execution. It does it by adding `input` to the executor environment, which is responsible for managing guest-readable memory. When the prover executes the program, it can access `input`:
```

    // Obtain the default prover.
    let prover = default_prover();
    // Produce a receipt by proving the specified ELF binary.
    let receipt = prover.prove_elf(env, HELLO_GUEST_ELF).unwrap();

    // Extract journal of receipt
    let _output: u32 = receipt.journal.decode().unwrap();

    receipt.verify(HELLO_GUEST_ID).unwrap();

    // Print, notice, after committing to a journal, the private input became public
    println!("Hello, world! I generated a proof of guest execution! {} is a public output from journal ", _output);
}
```

You should now be able to see your proof with `cargo run --release`.
If your program printed the "Hello, world!" assertion and [receipt] verification was a success, congratulations!
If not, we hope that troubleshooting will get you familiar with the system, and we'd love to chat with you on [Discord].
Or, if you believe you've found a bug or other problem in our code, please open an [issue] describing the problem.

If you're ready to start building more complex projects, we recommend taking a look at the other examples in our [examples directory] for more project ideas that use zero-knowledge proofs.

> [Installation]: https://dev.risczero.com/zkvm/install

Reviewer 2: We should avoid using absolute links for intra-site links

> [Installation]: /zkvm/install

Reviewer 2: ```suggestion
[Installation]: ../install
```

> [cargo risczero]: https://docs.rs/cargo-risczero/latest/cargo_risczero/

Reviewer 2: ```suggestion
[cargo risczero]: https://docs.rs/cargo-risczero
```
